File: js/src/server_warnings.test.ts
import { CallOptions, ClientError, Metadata, Status } from "nice-grpc";
import { expect, test, vi } from "vitest";
import type { Logger } from "./logger";
import {
  SERVER_WARNING_REGISTRY_LIMIT,
  logServerWarnings,
  serverWarningMiddleware,
} from "./server_warnings";

function makeLogger() {
  return {
    debug: vi.fn(),
    info: vi.fn(),
    warn: vi.fn(),
    error: vi.fn(),
  } satisfies Logger;
}

function makeCall(trailer: Metadata, err?: Error): any {
  return {
    method: {
      path: "/modal.client.ModalClient/AppGetOrCreate",
      requestStream: false,
      responseStream: false,
    },
    request: {},
    requestStream: false,
    responseStream: false,
    async *next(_request: unknown, options: CallOptions) {
      options.onTrailer?.(trailer);
      if (err) throw err;
      yield {};
    },
  };
}

async function drain(gen: AsyncIterable<unknown>) {
  for await (const _ of gen) {
    // Drain the response.
  }
}

test("serverWarningMiddleware logs decoded warnings", async () => {
  const logger = makeLogger();
  const trailer = new Metadata();
  trailer.append("x-modal-warning", "Hello%20world");
  trailer.append("x-modal-warning", "100%25%2C%20done");

  await drain(serverWarningMiddleware(logger)(makeCall(trailer), {}));

  expect(logger.warn.mock.calls).toEqual([["Hello world"], ["100%, done"]]);
});

test("serverWarningMiddleware logs warnings on failed calls", async () => {
  const logger = makeLogger();
  const trailer = new Metadata({ "x-modal-warning": "Something%20happened" });
  const rpcErr = new ClientError("/m", Status.INTERNAL, "boom");

  await expect(
    drain(serverWarningMiddleware(logger)(makeCall(trailer, rpcErr), {})),
  ).rejects.toBe(rpcErr);

  expect(logger.warn.mock.calls).toEqual([["Something happened"]]);
});

test("serverWarningMiddleware ignores other trailers", async () => {
  const logger = makeLogger();
  const trailer = new Metadata({ "x-other": "value" });

  await drain(serverWarningMiddleware(logger)(makeCall(trailer), {}));

  expect(logger.warn).not.toHaveBeenCalled();
});

test("serverWarningMiddleware logs malformed encodings verbatim", async () => {
  const logger = makeLogger();
  const trailer = new Metadata({ "x-modal-warning": "bad%E0%A4%A" });

  await drain(serverWarningMiddleware(logger)(makeCall(trailer), {}));

  expect(logger.warn.mock.calls).toEqual([["bad%E0%A4%A"]]);
});

test("serverWarningMiddleware forwards trailers to the caller", async () => {
  const logger = makeLogger();
  const trailer = new Metadata({ "x-modal-warning": "hi" });
  const onTrailer = vi.fn();

  await drain(
    serverWarningMiddleware(logger)(makeCall(trailer), { onTrailer }),
  );

  expect(onTrailer).toHaveBeenCalledWith(trailer);
});

test("serverWarningMiddleware logs a repeated warning once", async () => {
  const logger = makeLogger();
  const middleware = serverWarningMiddleware(logger);
  const trailer = new Metadata({ "x-modal-warning": "Token%20expires%20soon" });

  await drain(middleware(makeCall(trailer), {}));
  await drain(middleware(makeCall(trailer), {}));

  expect(logger.warn.mock.calls).toEqual([["Token expires soon"]]);
});

test("logServerWarnings keeps the registry bounded but never goes quiet", () => {
  // A server that varies the text of a warning grows the registry, but only up to the limit.
  const logger = makeLogger();
  const seen = new Set<string>();
  for (let i = 0; i < SERVER_WARNING_REGISTRY_LIMIT * 3; i++) {
    const trailer = new Metadata({
      "x-modal-warning": `warning about object ${i}`,
    });
    logServerWarnings(logger, trailer, seen);
  }
  // Overflow drops the entry for this message rather than silencing it.
  logServerWarnings(
    logger,
    new Metadata({ "x-modal-warning": "warning about object 0" }),
    seen,
  );

  expect(seen.size).toBeLessThanOrEqual(SERVER_WARNING_REGISTRY_LIMIT + 2);
  const messages = logger.warn.mock.calls.map(([m]) => m);
  expect(messages).toHaveLength(SERVER_WARNING_REGISTRY_LIMIT * 3 + 1);
  expect(messages.at(-1)).toBe("warning about object 0");
});

File: go/server_warnings_test.go
package modal

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"io"
	"log/slog"
	"strings"
	"testing"

	"github.com/onsi/gomega"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/metadata"
	"google.golang.org/grpc/status"
)

func newServerWarningTestLogger() (*serverWarningLogger, *bytes.Buffer) {
	var buf bytes.Buffer
	logger := slog.New(slog.NewTextHandler(&buf, &slog.HandlerOptions{Level: slog.LevelWarn}))
	return newServerWarningLogger(logger), &buf
}

func invokerWithTrailer(md metadata.MD, err error) grpc.UnaryInvoker {
	return func(ctx context.Context, method string, req, reply any, cc *grpc.ClientConn, opts ...grpc.CallOption) error {
		for _, o := range opts {
			if t, ok := o.(grpc.TrailerCallOption); ok {
				*t.TrailerAddr = md
			}
		}
		return err
	}
}

func TestServerWarningUnaryInterceptor(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	md := metadata.Pairs(serverWarningHeader, "Hello%20world", serverWarningHeader, "100%25%20done")
	err := serverWarningUnaryInterceptor(w)(t.Context(), "/m", nil, nil, nil, invokerWithTrailer(md, nil))
	g.Expect(err).ToNot(gomega.HaveOccurred())

	out := buf.String()
	g.Expect(out).To(gomega.ContainSubstring(`level=WARN msg="Hello world"`))
	g.Expect(out).To(gomega.ContainSubstring(`level=WARN msg="100% done"`))
}

func TestServerWarningUnaryInterceptorOnError(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	md := metadata.Pairs(serverWarningHeader, "Something%20happened")
	rpcErr := status.Error(codes.Internal, "boom")
	err := serverWarningUnaryInterceptor(w)(t.Context(), "/m", nil, nil, nil, invokerWithTrailer(md, rpcErr))
	g.Expect(err).To(gomega.Equal(rpcErr))
	g.Expect(buf.String()).To(gomega.ContainSubstring(`msg="Something happened"`))
}

func TestServerWarningUnaryInterceptorNoWarnings(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	md := metadata.Pairs("x-other", "value")
	err := serverWarningUnaryInterceptor(w)(t.Context(), "/m", nil, nil, nil, invokerWithTrailer(md, nil))
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(buf.String()).To(gomega.BeEmpty())
}

type fakeTrailerStream struct {
	grpc.ClientStream
	trailer metadata.MD
}

func (s *fakeTrailerStream) RecvMsg(any) error    { return io.EOF }
func (s *fakeTrailerStream) Trailer() metadata.MD { return s.trailer }

func TestServerWarningStreamInterceptor(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	fake := &fakeTrailerStream{trailer: metadata.Pairs(serverWarningHeader, "Stream%20warning")}
	streamer := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
		return fake, nil
	}
	cs, err := serverWarningStreamInterceptor(w)(t.Context(), &grpc.StreamDesc{ServerStreams: true}, nil, "/m", streamer)
	g.Expect(err).ToNot(gomega.HaveOccurred())

	g.Expect(errors.Is(cs.RecvMsg(nil), io.EOF)).To(gomega.BeTrue())
	g.Expect(errors.Is(cs.RecvMsg(nil), io.EOF)).To(gomega.BeTrue())
	g.Expect(strings.Count(buf.String(), `msg="Stream warning"`)).To(gomega.Equal(1))
}

func TestServerWarningRepeatsSuppressed(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	same := metadata.Pairs(serverWarningHeader, "Same%20warning")
	for range 5 {
		err := serverWarningUnaryInterceptor(w)(t.Context(), "/m", nil, nil, nil, invokerWithTrailer(same, nil))
		g.Expect(err).ToNot(gomega.HaveOccurred())
	}
	other := metadata.Pairs(serverWarningHeader, "Other%20warning")
	err := serverWarningUnaryInterceptor(w)(t.Context(), "/m", nil, nil, nil, invokerWithTrailer(other, nil))
	g.Expect(err).ToNot(gomega.HaveOccurred())

	out := buf.String()
	g.Expect(strings.Count(out, `msg="Same warning"`)).To(gomega.Equal(1))
	g.Expect(strings.Count(out, `msg="Other warning"`)).To(gomega.Equal(1))
}

func TestServerWarningRepeatsSuppressedAcrossStreams(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	w, buf := newServerWarningTestLogger()
	streamer := func(ctx context.Context, desc *grpc.StreamDesc, cc *grpc.ClientConn, method string, opts ...grpc.CallOption) (grpc.ClientStream, error) {
		return &fakeTrailerStream{trailer: metadata.Pairs(serverWarningHeader, "Stream%20warning")}, nil
	}
	for range 2 {
		cs, err := serverWarningStreamInterceptor(w)(t.Context(), &grpc.StreamDesc{ServerStreams: true}, nil, "/m", streamer)
		g.Expect(err).ToNot(gomega.HaveOccurred())
		g.Expect(errors.Is(cs.RecvMsg(nil), io.EOF)).To(gomega.BeTrue())
	}
	g.Expect(strings.Count(buf.String(), `msg="Stream warning"`)).To(gomega.Equal(1))
}

func TestServerWarningRegistryBounded(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)

	var r serverWarningRegistry
	logged := 0
	for i := range serverWarningRegistryLimit * 3 {
		if r.shouldLog(fmt.Sprintf("warning about object %d", i)) {
			logged++
		}
	}
	g.Expect(logged).To(gomega.Equal(serverWarningRegistryLimit * 3))
	g.Expect(len(r.seen)).To(gomega.BeNumerically("<=", serverWarningRegistryLimit))
	// Overflow drops the entry for this message rather than silencing it.
	g.Expect(r.shouldLog("warning about object 0")).To(gomega.BeTrue())
}
